Author : [ Snorre Johnsen ]
Course : Numerical Scientific Computing 2026

# Milestone 1 - function level profiling

In [5]:
import numpy as np 
import cProfile, pstats
from mandelbrot import compute_mandelbrot_naive, compute_mandelbrot_vectorized


In [7]:
cProfile.run('compute_mandelbrot_naive(-2, 1, -1.5, 1.5, 1024, 1024)', 'naive_profile.prof')
cProfile.run('compute_mandelbrot_vectorized(-2, 1, -1.5, 1.5, 1024, 1024)', 'numpy_profile.prof')

for name in ('naive_profile.prof', 'numpy_profile.prof'):
    stats = pstats.Stats(name)
    stats.sort_stats('cumulative')
    stats.print_stats(10)


Thu Feb 26 14:00:02 2026    naive_profile.prof

         23008341 function calls in 13.179 seconds

   Ordered by: cumulative time
   List reduced from 20 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   13.179   13.179 {built-in method builtins.exec}
        1    0.000    0.000   13.179   13.179 <string>:1(<module>)
        1    3.121    3.121   13.179   13.179 /home/snorre/Desktop/es8/courses/nsc-course/mandelbrot-nsc/mandelbrot.py:50(compute_mandelbrot_naive)
  1048576    7.900    0.000   10.056    0.000 /home/snorre/Desktop/es8/courses/nsc-course/mandelbrot-nsc/mandelbrot.py:25(mandelbrot_point_naive)
 21959734    2.156    0.000    2.156    0.000 {built-in method builtins.abs}
        2    0.000    0.000    0.001    0.000 /home/snorre/miniforge3/envs/nsc2026/lib/python3.11/site-packages/numpy/_core/function_base.py:26(linspace)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.zer

#### Which function takes most total time?
for the naive impolementaion it is the mandelbrot_point function.

i cannot see the numpy profiling it seems as the cprofiler cannot do it on numpy.
#### Are there functions called surprisingly many times?
the python abs() function is called alot of times (21959734 times)
#### How does the NumPy profile compare to naive?
as mentioned before cProfiler cannot seem to read numpy but the numpy is significantly faster

# Milestone 2

tried both with @profile and line_profiler.profile

seems to get roughly same result